server/controllers.go

```diff
@@ -5,192 +5,192 @@ import (
 	"embed"
 	"fmt"
 	"io/fs"
+	"time"
 
-	filteredcache "github.com/IBM/controller-filtered-cache/filteredcache"
 	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
-	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
 	genericapiserver "k8s.io/apiserver/pkg/server"
 	"k8s.io/client-go/dynamic"
-	"k8s.io/client-go/rest"
-	policyv1 "open-cluster-management.io/governance-policy-propagator/api/v1"
-	placementrulev1 "open-cluster-management.io/multicloud-operators-subscription/pkg/apis/apps/placementrule/v1"
-	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/yaml"
 
 	"github.com/clyang82/multicluster-global-hub-lite/server/controllers"
 )
 
-const (
-// rootPolicyLabel    = "policy.open-cluster-management.io/root-policy"
-// localResourceLabel = "multicluster-global-hub.open-cluster-management.io/local-resource"
-)
-
 //go:embed manifests
 var crdManifestsFS embed.FS
 
-func (s *GlobalHubApiServer) CreateCache(ctx context.Context, cfg *rest.Config) error {
+func (s *GlobalHubApiServer) installCRD(ctx context.Context) error {
+	controllerName := "global-hub-crd-controller"
+	s.AddPostStartHook(fmt.Sprintf("start-%s", controllerName),
+		func(hookContext genericapiserver.PostStartHookContext) error {
+			fs.WalkDir(crdManifestsFS, "manifests", func(file string, d fs.DirEntry, err error) error {
+				if err != nil {
+					return err
+				}
+
+				if !d.IsDir() {
+					b, err := crdManifestsFS.ReadFile(file)
+					if err != nil {
+						return err
+					}
+					obj := &unstructured.Unstructured{}
+					err = yaml.Unmarshal(b, &obj)
+					if err != nil {
+						return err
+					}
+					_, err = s.client.
+						Resource(apiextensionsv1.SchemeGroupVersion.WithResource("customresourcedefinitions")).
+						Create(context.TODO(), obj, metav1.CreateOptions{})
+					if err != nil {
+						return err
+					}
+				}
+				return nil
+			})
+			return nil
+		})
+	return nil
+}
+
+func (s *GlobalHubApiServer) GetClient() dynamic.Interface {
+	return s.client
+}
 
-	scheme := runtime.NewScheme()
+type globalHubController struct {
+	name           string
+	ctx            context.Context
+	gvr            schema.GroupVersionResource
+	informer       cache.SharedIndexInformer
+	queue          workqueue.RateLimitingInterface
+	createInstance func() client.Object
+	client         dynamic.Interface
+	reconcile      func(ctx context.Context, obj interface{}) error
+}
```

Comment:
I prefer to put this to package `github.com/clyang82/multicluster-global-hub-lite/server/controllers`

Reply:
Your advice is great! I have combined the two controller file together.